### ISSUE TITLE
Allow Server SNI Interrogation in Transparent Mode

### ISSUE THREAD

**Issue by ericdraken** [mitmproxy/mitmproxy#5064]
#### Problem Description

The sever SNI is not added to the array of hostnames checked against `ignored-hosts` and `allowed-hosts` in `next_layer.py` so some requests get past the respective filters.

#### Proposal

Here is a patch I currently use myself that I'd like to share (7.0.4):

```diff
Index: venv/lib/python3.8/site-packages/mitmproxy/addons/next_layer.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/usr/local/lib/python3.8/site-packages/mitmproxy/addons/next_layer.py b/usr/local/lib/python3.8/site-packages/mitmproxy/addons/next_layer.py
--- a/usr/local/lib/python3.8/site-packages/mitmproxy/addons/next_layer.py  (date 1641187083049)
+++ b/usr/local/lib/python3.8/site-packages/mitmproxy/addons/next_layer.py  (date 1641187083049)
@@ -59,7 +59,7 @@
                 re.compile(x, re.IGNORECASE) for x in ctx.options.allow_hosts
             ]
 
-    def ignore_connection(self, server_address: Optional[connection.Address], data_client: bytes) -> Optional[bool]:
+    def ignore_connection(self, server: Optional[connection.Server], data_client: bytes) -> Optional[bool]:
         """
         Returns:
             True, if the connection should be ignored.
@@ -70,8 +70,11 @@
             return False
 
         hostnames: List[str] = []
-        if server_address is not None:
-            hostnames.append(server_address[0])
+        if server is not None:
+            if server.address is not None:
+                hostnames.append(server.address[0])
+            if server.sni is not None:
+                hostnames.append(server.sni)
         if is_tls_record_magic(data_client):
             try:
                 ch = parse_client_hello(data_client)
@@ -122,7 +125,7 @@
             return stack_match(context, layers)
 
         # 1. check for --ignore/--allow
-        ignore = self.ignore_connection(context.server.address, data_client)
+        ignore = self.ignore_connection(context.server, data_client)
         if ignore is True:
             return layers.TCPLayer(context, ignore=True)
         if ignore is None:
```

#### Alternatives
N/A

#### Additional context

Here is a write-up I made about why this patch is needed: https://ericdraken.com/pfsense-decrypt-ad-traffic/#patch

**Comment by mhils** [comment]
Thanks for sharing your patches! 😃 
Could you share how your setup looks like? We do peek at the incoming bytes and extract the SNI right after your additions:
https://github.com/mitmproxy/mitmproxy/blob/1abb8f69217910c8623bd1339da2502aed98ff0d/mitmproxy/addons/next_layer.py#L75-L85

If that for some reason doesn't pick up a SNI that's a bug we should fix.

**Comment by ericdraken** [comment]
My pleasure. Well, I wrote an adblocker script and noticed that sometimes in the `hostnames` list there would only be an IP address, and sometimes an IP address and an `*.apple.com` FQDN. I was wondering why only an IP sometimes. I looked at the snipped you shared early on and when I breakpoint I often see just an IP.

**Here is the thing:** everything works perfectly in proxy mode or regular mode (I get FQDN each time). The problem I described above happens in transparent mode. 

I noticed that the `server.sni` field is available but not appended, and to my delight when I applied my patch, I faithfully see the FQDN of each request. I may need an explanation why it works now as I followed a hunch and got lucky(?).

**Comment by mhils** [comment]
> The problem I described above happens in transparent mode.

This is somewhat confusing to me - `client.sni` should only be set when the connection is already being intercepted, so your allow/ignore would come too late to stop that. The correct way is really to check if the client sends a client hello (`is_tls_record_magic`) and then parsing said ClientHello. If you have a case where that logic fails I'd be eager to see more, because that then would be a bug we should track down.

**Comment by ericdraken** [comment]
I hope this helps. I put a simple `print(hostnames)` at 

https://github.com/mitmproxy/mitmproxy/blob/1abb8f69217910c8623bd1339da2502aed98ff0d/mitmproxy/addons/next_layer.py#L89

Without my patch:

![image](https://user-images.githubusercontent.com/22000765/149736930-117e5523-7dd8-43c1-8736-cf10f3e61b5c.png)

With my patch:

![image](https://user-images.githubusercontent.com/22000765/149737428-09231609-4de8-4189-8f1e-6f66aa3badc7.png)

Right after each of the arrays is printed, the `ignore_connection` method evaluates an `any` check on the hostname. Unless I'm not understanding correctly, I'd prefer that each connection is rejected or accepted, not just the very first one (screenshot 1).

If it helps, I used `--no-http2` religiously. ;)

**Comment by ericdraken** [comment]
Thank you for indulging me in this conversation.

**Update:** I believe the problem relates to how I update the `NextLayer` regex.

Here is a minimal addon demonstrating the problem. I'd like to update the `allow-hosts` or `ignore-hosts` regex through code without triggering the update chain for all addons (this could be the mistake).

```
# -*- coding: utf-8 -*-
#  Copyright (c) 2021. Eric Draken (ericdraken.com)
#  PoC for Issue #5064
import re

from mitmproxy import ctx, http
from mitmproxy.addons.next_layer import NextLayer
from mitmproxy.proxy import layer

logger = ctx.log


class MITMLogger:
    """PoC for Issue #5064"""

    # Intercept only these wildcard domains and no others
    intercept_hosts = r"example\.(com|net|org)"
    intercept_hosts_re = re.compile(intercept_hosts, re.IGNORECASE)

    """The following methods are hooks that are called during the normal flow of MITMProxy"""

    # noinspection PyMethodMayBeStatic
    def load(self, _):
        # Do not allow piped requests we could miss
        ctx.options.http2 = False
        try:
            ctx.options.update(http2=False, anticomp=True, mode="transparent", termlog_verbosity="debug")
        except KeyError:
            ctx.options.update(http2=False, anticomp=True, mode="transparent")
        logger.warn(f"{self.__class__.__name__} loaded")

    def running(self):
        """Override CLI arguments and update the regex manually"""
        if ctx.options.allow_hosts != [self.intercept_hosts]:
            ctx.options.allow_hosts = [self.intercept_hosts]

            ctx.options.ignore_hosts = []
            next_layer_addon: NextLayer = ctx.master.addons.get("NextLayer".lower())
            next_layer_addon.configure("allow_hosts")
            logger.warn(f"Updated interceptable hosts")

    def next_layer(self, nextlayer: layer.NextLayer):
        """Do I want to passthru or block the request? Here is my chance."""
        logger.warn(nextlayer) # Usually none, so repr() should be changed in NextLayer
        logger.warn(nextlayer.context)

    def request(self, flow: http.HTTPFlow) -> None:
        logger.warn(f">> {flow.request.pretty_url}\n\n")

    # noinspection PyMethodMayBeStatic
    def error(self, flow: http.HTTPFlow):
        logger.warn(f">> {flow.request.pretty_url}\n<< {flow.error}\n\n")

    def response(self, flow: http.HTTPFlow):
        logger.warn(f">> {flow.request.pretty_url}\n<< {flow.response}\n\n")


# Register the addon
addons = [MITMLogger()]
```

Simple curl:

`curl --ssl-no-revoke https://example.com/`

![image](https://user-images.githubusercontent.com/22000765/149833856-fb00c6b8-fe5e-447e-b751-ca6c89399826.png)

![image](https://user-images.githubusercontent.com/22000765/149833869-fb402efc-7a0e-4630-a0c0-0921740d7101.png)

However, if adding `--allow-hosts` the normal way:

![image](https://user-images.githubusercontent.com/22000765/149833737-17c727ad-e9bf-4758-84fb-d28bb5659028.png)

![image](https://user-images.githubusercontent.com/22000765/149833744-6001c05e-3fc0-4b3e-a7bc-62e25ef511bb.png)

Somehow, I get both the server IP and hostname with my patch + setting the regex manually. The root of this issue may be in the way I update the regex. A better approach? I need to update the regex frequently.

---

> Could you please try the following print statment instead?
> print(f"{hostnames=} {is_tls_record_magic(data_client)=} {data_client=} {server=}") 

Certainly. Using the regex update:

![image](https://user-images.githubusercontent.com/22000765/149834446-08c5dff9-fcb1-49e7-8612-1ca4103b7c32.png)

**Comment by gotoin** [comment]
+1

**Comment by JarLob** [comment]
Hi @mhils, hopefully I can contribute something back to this amazing tool!
As you know I'm using mitmproxy in [permissions monitor](https://github.com/GitHubSecurityLab/actions-permissions). Recently I was experimenting with `allow_hosts` and noticed the same inconsistency when mitmproxy doesn't recognize some requests. I have forked mitmproxy, added some logging and hacked with the fork of the actions-permissions to get [this log](https://github.com/JarLob/proxy/actions/runs/9813122161/job/27098521689#step:12:122). A copy - [job-logs.txt](https://github.com/user-attachments/files/16113537/job-logs.txt)

![Image](https://github.com/user-attachments/assets/ec360331-3bb5-40e2-ba04-2fcd3edfd6a4)

As the author of this thread I see a tcp request that is not recognized as tls, but the `context.server.sni` has the host already.

Everything is in GitHub and should be reproducible. There is a fork of mitmproxy with some logging added (branch `sni`) - the [snapshot](https://github.com/JarLob/mitmproxy/tree/c6edb94023b8a27a4a76a9b8174f8cc1ec231b79). There is [fork](https://github.com/GitHubSecurityLab/actions-permissions/tree/maven) with hacks of permissions monitor action and there is [repository](https://github.com/JarLob/proxy/tree/sni) I use to start a workflow ([test.yml](https://github.com/JarLob/proxy/blob/fe4f9e496001952bcc7cf4b46b56b8d4b4fb3e20/.github/workflows/test.yml)) that calls curl with mitmproxy set up. But it basically boils down to this:
* Mitmproxy runs in transparent mode:
```
/home/mitmproxyuser/mitmproxy/venv/bin/mitmdump --mode transparent \
          --showhost \
          --allow-hosts '\bgithub\.com(:\d+)$' \
          --set block_global=false \
          `#-q` \
          --set termlog_verbosity=debug \
          --set proxy_debug=true \
          -s /home/mitmproxyuser/mitm_plugin.py \
          --set output='/home/mitmproxyuser/out.txt' \
          --set token='$INPUT_TOKEN' \
          --set hosts='api.github.com,github.com' \
          --set debug='$RUNNER_DEBUG' \
          --set ACTIONS_ID_TOKEN_REQUEST_URL='$ACTIONS_ID_TOKEN_REQUEST_URL' \
          --set ACTIONS_ID_TOKEN_REQUEST_TOKEN='$ACTIONS_ID_TOKEN_REQUEST_TOKEN' \
          --set GITHUB_REPOSITORY_ID='$GITHUB_REPOSITORY_ID' \
          --set GITHUB_REPOSITORY='$GITHUB_REPOSITORY' \
          >>/home/mitmproxyuser/out.txt 2>&1 &"
```
* Actions runner does some source checkout (this is why there is one correct interception) and then this curl command is run:
```
curl \
            -H "Accept: application/vnd.github+json" \
            -H "Authorization: Bearer ${{ github.token }}"\
            -H "X-GitHub-Api-Version: 2022-11-28" \
            https://Api.github.com/repos/${{ github.event.repository.owner.login }}/${{ github.event.repository.name }}/issues/1/labels
```
Everything else is in the logs, but I find it hard to understand :/ Don't hesitate to fork and run or guide me what else logging to add and I can run it for you.

P.S. I have noticed [these lines](https://github.com/JarLob/mitmproxy/blob/c6edb94023b8a27a4a76a9b8174f8cc1ec231b79/mitmproxy/addons/next_layer.py#L241-L244) were not under `if context.server.address:` in previous versions of mitmproxy and shifted them left, but it did seem to help. What version do you think is correct?

**Comment by mhils** [comment]
Thank you @JarLob! Your logs were immensely helpful. 😃 This shoulf be fixed in #7002, see the PR for a detailed description. :)
